# Анализ данных SQL по купленному сервису для чтения книг по подписке

<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Описание-данных" data-toc-modified-id="Описание-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Описание данных</a></span></li><li><span><a href="#Исследование-таблиц" data-toc-modified-id="Исследование-таблиц-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Исследование таблиц</a></span></li><li><span><a href="#Анализ-данных" data-toc-modified-id="Анализ-данных-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Анализ данных</a></span><ul class="toc-item"><li><span><a href="#Посчитаем,-сколько-книг-вышло-после-1-января-2000-года" data-toc-modified-id="Посчитаем,-сколько-книг-вышло-после-1-января-2000-года-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Посчитаем, сколько книг вышло после 1 января 2000 года</a></span></li><li><span><a href="#Для-каждой-книги-посчитаем-количество-обзоров-и-среднюю-оценку" data-toc-modified-id="Для-каждой-книги-посчитаем-количество-обзоров-и-среднюю-оценку-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Для каждой книги посчитаем количество обзоров и среднюю оценку</a></span></li><li><span><a href="#Определим-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-—-так-вы-исключите-из-анализа-брошюры" data-toc-modified-id="Определим-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-—-так-вы-исключите-из-анализа-брошюры-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Определим издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры</a></span></li><li><span><a href="#Определим-автора-с-самой-высокой-средней-оценкой-книг-—-учитывайте-только-книги-с-50-и-более-оценками" data-toc-modified-id="Определим-автора-с-самой-высокой-средней-оценкой-книг-—-учитывайте-только-книги-с-50-и-более-оценками-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Определим автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками</a></span></li><li><span><a href="#Посчитаем-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-48-оценок." data-toc-modified-id="Посчитаем-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-48-оценок.-3.5"><span class="toc-item-num">3.5&nbsp;&nbsp;</span>Посчитаем среднее количество обзоров от пользователей, которые поставили больше 48 оценок.</a></span></li></ul></li><li><span><a href="#Выводы" data-toc-modified-id="Выводы-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Выводы</a></span></li></ul></div>

## Описание данных 

**Цели исследования**

Проанализировать данные купленного сервиса для чтения книг по подписке для помощи в формировании ценностного предложения нового продукта.

**Задачи** 

1. Подготовить анализ:

* ĸоличества ĸниг,

* обзоров, 

* рейтингов 

2. Сформулировать выводы

**Описание проекта**
Задача  — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

**Задания**

* Посчитайте, сколько книг вышло после 1 января 2000 года;

* Для каждой книги посчитайте количество обзоров и среднюю оценку;

* Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;

* Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;

* Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

**Ход работы** 

* Опишите цели исследования;

* Исследуйте таблицы — выведите первые строки;

* Сделайте по одному SQL-запросу для решения каждого задания;

* Выведите результат каждого запроса в тетрадке;

* Опишите выводы по каждой из решённых задач.

## Исследование таблиц

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import text, create_engine
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
    'pwd': 'Sdf4$2;d-d30pp', # пароль
    'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
    'port': 6432, # порт подключения
    'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})
# чтобы выполнить SQL-запрос, используем Pandas
query = '''SELECT * FROM books LIMIT 5'''
# открываем соединение, получаем данные
with engine.connect() as con:
    df = pd.io.sql.read_sql(sql=text(query), con = con)
# выведем данные    
df

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

Прочитаем данные и выведем на экран первые 5 строчек датафрейма books.

In [2]:
query = '''
            SELECT * 
            FROM books
        '''
books = pd.io.sql.read_sql(query, con = engine)
books.head()

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

Посмотрим общие данные, выведя основную информацию о датафрейме с помощью метода info().

In [3]:
books.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


Датафрейм books содержит 1000 строки и 6 столбца.

Прочитаем данные и выведем на экран первые 5 строчек датафрейма authors.

In [4]:
query = '''
            SELECT * 
            FROM authors
        '''
authors = pd.io.sql.read_sql(query, con = engine)
authors.head()

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

Посмотрим общие данные, выведя основную информацию о датафрейме с помощью метода info().

In [5]:
authors.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


Датафрейм authors содержит 636 строки и 2 столбца.

Прочитаем данные и выведем на экран первые 5 строчек датафрейма publishers.

In [6]:
query = '''
            SELECT * 
            FROM publishers
        '''
publishers = pd.io.sql.read_sql(query, con = engine)
publishers.head()

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

Посмотрим общие данные, выведя основную информацию о датафрейме с помощью метода info().

In [7]:
publishers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


Датафрейм publisher содержит 340 строки и 2 столбца.

Прочитаем данные и выведем на экран первые 5 строчек датафрейма ratings.

In [8]:
query = '''
            SELECT * 
            FROM ratings
        '''
ratings = pd.io.sql.read_sql(query, con = engine)
ratings.head()

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

Посмотрим общие данные, выведя основную информацию о датафрейме с помощью метода info().

In [9]:
ratings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


Датафрейм ratings содержит 6456 строки и 4 столбца.

Прочитаем данные и выведем на экран первые 5 строчек датафрейма reviews.

In [10]:
query = '''
            SELECT * 
            FROM reviews
        '''
reviews = pd.io.sql.read_sql(query, con = engine)
reviews.head()

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

Посмотрим общие данные, выведя основную информацию о датафрейме с помощью метода info().

In [11]:
reviews.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


Датафрейм reviews содержит 2793 строки и 4 столбца.

**Вывод**: Вывели данные, изучили датафреймы и посмотрели общие данные. В данных отсутствуют пропуски, тип данных не требует изменений.

## Анализ данных

### Посчитаем, сколько книг вышло после 1 января 2000 года

In [12]:
query = '''
            SELECT COUNT(book_id) AS count_books
            FROM books 
            WHERE publication_date > '01-01-2000'
        '''
pd.io.sql.read_sql(query, con = engine)

count_books
0          819

**Вывод**: 819 книг вышло после 1 января 2000 года.

### Для каждой книги посчитаем количество обзоров и среднюю оценку

In [13]:
query = '''
            SELECT books.title AS title, 
                   books.book_id AS id, 
                   AVG(ratings.rating) AS avg_rating, 
                   COUNT(DISTINCT reviews.review_id) AS count_reviews
            FROM books
            LEFT JOIN ratings ON ratings.book_id = books.book_id
            LEFT JOIN reviews ON reviews.book_id = books.book_id
            GROUP BY title, 
                     books.book_id
            ORDER BY count_reviews DESC
            
        '''
pd.io.sql.read_sql(query, con = engine)

title   id  avg_rating  \
0                              Twilight (Twilight  #1)  948    3.662500   
1                                  Water for Elephants  963    3.977273   
2                                     The Glass Castle  734    4.206897   
3    Harry Potter and the Prisoner of Azkaban (Harr...  302    4.414634   
4    The Curious Incident of the Dog in the Night-Time  695    4.081081   
..                                                 ...  ...         ...   
995    Anne Rice's The Vampire Lestat: A Graphic Novel   83    3.666667   
996                            The Natural Way to Draw  808    3.000000   
997   The Cat in the Hat and Other Dr. Seuss Favorites  672    5.000000   
998                          Essential Tales and Poems  221    4.000000   
999  Disney's Beauty and the Beast (A Little Golden...  191    4.000000   

     count_reviews  
0                7  
1                6  
2                6  
3                6  
4                6  
..             ...  
995              0  
996              0  
997              0  
998              0  
999              0  

[1000 rows x 4 columns]

**Вывод**: Самое большое количество обзоров занимает книга Twilight (Twilight #1). Количество обзоров равняется 7. Средняя оценка 3.66. Очень много книг со средней оценкой 5, но количество обзоров в основном равняется 2.

### Определим издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры

In [14]:
query = '''
            SELECT COUNT(books.book_id) AS count_books,
                   publishers.publisher
            FROM publishers
            INNER JOIN books ON books.publisher_id = publishers.publisher_id
            WHERE num_pages > 50
            GROUP BY publishers.publisher_id
            ORDER BY count_books DESC
            LIMIT 20;
        '''
pd.io.sql.read_sql(query, con = engine)

count_books                  publisher
0            42              Penguin Books
1            31                    Vintage
2            25   Grand Central Publishing
3            24           Penguin Classics
4            19           Ballantine Books
5            19                     Bantam
6            17                    Berkley
7            14              Berkley Books
8            14         St. Martin's Press
9            13                      Delta
10           13  William Morrow Paperbacks
11           12              HarperCollins
12           12  Little  Brown and Company
13           12           Harper Perennial
14           11                  Tor Books
15           11             Back Bay Books
16           10                     Signet
17           10                    Vertigo
18           10              Mariner Books
19           10                    Penguin

**Вывод**: Издательство, которое выпустило наибольшее число книг толще 50 страниц — Penguin Books. Число выпущенных книг составляет 42.

### Определим автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками

In [15]:
query = '''
            SELECT i.author AS author,
                   AVG(i.avg_rating) AS avg_rating
            FROM 
                (SELECT a.author AS author,
                        b.book_id AS id,
                        AVG(r.rating) AS avg_rating,
                        COUNT(r.rating) AS count_rating
                FROM authors AS a
                INNER JOIN books AS b ON b.author_id = a.author_id
                INNER JOIN ratings AS r ON r.book_id = b.book_id
                GROUP BY a.author, b.book_id
                HAVING COUNT(r.rating) >= 50) AS i
                
            GROUP BY author
            ORDER BY avg_rating DESC
        '''
pd.io.sql.read_sql(query, con = engine)

author  avg_rating
0                          J.K. Rowling/Mary GrandPré    4.283844
1                   Markus Zusak/Cao Xuân Việt Khương    4.264151
2                                      J.R.R. Tolkien    4.258446
3                                   Louisa May Alcott    4.192308
4                                        Rick Riordan    4.080645
5                                     William Golding    3.901408
6                                       J.D. Salinger    3.825581
7            Paulo Coelho/Alan R. Clarke/Özdemir İnce    3.789474
8   William Shakespeare/Paul Werstine/Barbara A. M...    3.787879
9                                           Dan Brown    3.754540
10                                         Lois Lowry    3.750000
11            George Orwell/Boris Grabnar/Peter Škerl    3.729730
12                                    Stephenie Meyer    3.662500
13                                     John Steinbeck    3.622951

**Вывод**: Автор с самой высокой средней оценкой книг(учитывая только книги с 50 и более оценками) — J.K. Rowling/Mary GrandPré. Средняя оценка его книг составляет 4.28.

### Посчитаем среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

In [16]:
query = '''
            SELECT AVG(i.count_reviews) AS avg_count_review
            FROM 
                (SELECT username,
                        COUNT(review_id) AS count_reviews
                FROM reviews
                WHERE username IN
                        (SELECT username
                        FROM ratings
                        GROUP BY username
                        HAVING COUNT(rating_id) > 48)
                GROUP BY username) AS i
        '''
pd.io.sql.read_sql(query, con = engine)

avg_count_review
0              24.0

**Вывод**: Среднее количество обзоров от пользователей, которые поставили больше 48 оценок равняется 24.

## Выводы

* 819 книг вышло после 1 января 2000 года,

* Самое большое количество обзоров занимает книга Twilight (Twilight #1). Количество обзоров равняется 7. Средняя оценка 3.66. Очень много книг со средней оценкой 5, но количество обзоров в основном равняется 2.

* Издательство, которое выпустило наибольшее число книг толще 50 страниц — Penguin Books. Число выпущенных книг составляет 42.

* Автор с самой высокой средней оценкой книг(учитывая только книги с 50 и более оценками) — J.K. Rowling/Mary GrandPré. Средняя оценка его книг составляет 4.28.

* Среднее количество обзоров от пользователей, которые поставили больше 48 оценок равняется 24.